pkg/subctl/cmd/gather.go

```diff
@@ -0,0 +1,170 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/submariner-operator/pkg/internal/cli"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/components"
+)
+
+var (
+	gatherType   string
+	gatherModule string
+)
+
+const (
+	Logs = "logs"
+	Resources = "resources"
+)
+var gatherModuleFlags = map[string]bool{
+	components.Connectivity:     false,
+	components.ServiceDiscovery: false,
+	components.Broker:           false,
+	components.Operator:         false,
+}
+
+var gatherTypeFlags = map[string]bool{
+	"logs":      false,
+	"resources": false,
+}
+
+var gatherFuncs = map[string]interface{}{
+	components.Connectivity:     gatherConnectivity,
+	components.ServiceDiscovery: gatherDiscovery,
+	components.Broker:           gatherBroker,
+	components.Operator:         gatherOperator,
+}
+
+func init() {
+	addKubeconfigFlag(gatherCmd)
+	addGatherFlags(gatherCmd)
+	rootCmd.AddCommand(gatherCmd)
+}
+
+func addGatherFlags(gatherCmd *cobra.Command) {
+	gatherCmd.Flags().StringVar(&gatherType, "type", strings.Join(getAllTypeKeys(), ","), "comma-separated list of data types to gather")
+	gatherCmd.Flags().StringVar(&gatherModule, "module", strings.Join(getAllModuleKeys(), ","), "comma-separated list of components for which to gather data")
+}
+
+var gatherCmd = &cobra.Command{
+	Use:   "gather <kubeConfig>",
+	Short: "Gather troubleshooting data from a cluster",
+	Long: fmt.Sprintf("This command gathers data from a submariner cluster for troubleshooting. Data gathered"+
+		"can be selected by component (%v) and type (%v). Default is to capture all data.", strings.Join(getAllModuleKeys(), ","), strings.Join(getAllTypeKeys(), ",")),
+	Run: func(cmd *cobra.Command, args []string) {
+		gatherData()
+	},
+}
+
+func gatherData() {
+	err := checkGatherArguments()
+	exitOnError("Invalid arguments", err)
+
+	status := cli.NewStatus()
+	status.Start("Gathering Data...")
+	for module, ok := range gatherModuleFlags {
+		if ok {
+			for dataType, ok := range gatherTypeFlags {
+				if ok {
+					err = gatherFuncs[module].(func(*cli.Status, string) error)(status, dataType)
+				}
+			}
+		}
+	}
+	status.End(cli.CheckForError(err))
+	exitOnError("Error deploying the operator", err)
```

Comment:
I think it would be better to create a status spinner per module/type. This way we can pass/fail each one.

```suggestion
	for module, ok := range gatherModuleFlags {
		if ok {
			for dataType, ok := range gatherTypeFlags {
				if ok {
					status := cli.NewStatus()
					status.Start(fmt.Sprintf("Gathering %s %s...", module, dataType))
					status.End(cli.CheckForError(gatherFuncs[module](status, dataType)))
				}
			}
		}
	}
```

Reply:
Yeah, I was in two minds on what to go for. Went with single one coz not sure we want to use a common `cli.Status` across all. Note that it is declared as a global in `cmd`. But that needs some rework, and can revisit if decide to go that way.

Reply:
You're using a local `Status` here either here. When I was playing around with it, it seemed better to me to have a spinner/pass/fail marker next to each one.

Reply:
Good point, I agree having separate spinners would be better.